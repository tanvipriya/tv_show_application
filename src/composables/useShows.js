import { ref } from 'vue';
import axios from 'axios';

export function useShows() {
    const shows = ref([]);
    const loading = ref(false);
    const error = ref(null);

    /**
     * Fetch TV shows from the API
     * @param {number} page - Page number to fetch (default: 0)
     */
    const fetchShows = async (page = 0) => {
        loading.value = true;
        error.value = null;

        try {
            const { data } = await axios.get(`https://api.tvmaze.com/shows`, {
                params: { page }
            });
            shows.value = data;
        } catch (err) {
            error.value = err?.response?.data || err.message || 'Unknown error';
            console.error('Error fetching shows:', error.value);
        } finally {
            loading.value = false;
        }
    };

    /**
     * Clear the shows list
     */
    const clearShows = () => {
        shows.value = [];
    };

    return {
        shows,
        loading,
        error,
        fetchShows,
        clearShows
    };
}
