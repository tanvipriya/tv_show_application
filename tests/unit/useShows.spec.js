import { ref } from 'vue';
import axios from 'axios';
import { useShows } from '@/composables/useShows'; // adjust the import path as needed

jest.mock('axios');

describe('useShows composable with snapshots', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('initializes with default values', () => {
        const { shows, loading, error } = useShows();

        // Snapshot of initial state
        expect({ shows: shows.value, loading: loading.value, error: error.value }).toMatchSnapshot();
    });

    it('fetches shows successfully', async () => {
        const mockData = [
            { id: 1, name: 'Breaking Bad' },
            { id: 2, name: 'Friends' },
        ];
        axios.get.mockResolvedValueOnce({ data: mockData });

        const { shows, loading, error, fetchShows } = useShows();

        // Trigger fetch
        const promise = fetchShows(0);

        // Snapshot immediately after calling fetchShows (loading should be true)
        expect({ shows: shows.value, loading: loading.value, error: error.value }).toMatchSnapshot('loading state');

        await promise; // wait for fetch to resolve

        // Snapshot after successful fetch
        expect({ shows: shows.value, loading: loading.value, error: error.value }).toMatchSnapshot('after successful fetch');
    });

    it('handles API errors correctly', async () => {
        const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => { });

        const mockError = { response: { data: 'API Error Message' } };
        axios.get.mockRejectedValueOnce(mockError);

        const { shows, loading, error, fetchShows } = useShows();
        await fetchShows(1);

        expect({ shows: shows.value, loading: loading.value, error: error.value }).toMatchSnapshot();

        consoleSpy.mockRestore();
    });

    it('handles unknown errors gracefully', async () => {
        const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => { });

        axios.get.mockRejectedValueOnce(new Error('Network issue'));

        const { error, fetchShows } = useShows();
        await fetchShows();

        expect({ error: error.value }).toMatchSnapshot();

        consoleSpy.mockRestore();
    });


    it('clears shows list', () => {
        const { shows, clearShows } = useShows();
        shows.value = [{ id: 1, name: 'Test Show' }];

        clearShows();

        // Snapshot after clearing shows
        expect({ shows: shows.value }).toMatchSnapshot('after clearing shows');
    });
});
