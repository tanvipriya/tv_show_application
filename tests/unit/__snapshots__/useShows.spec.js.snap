// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useShows composable with snapshots clears shows list: after clearing shows 1`] = `
Object {
  "shows": Array [],
}
`;

exports[`useShows composable with snapshots fetches shows successfully: after successful fetch 1`] = `
Object {
  "error": null,
  "loading": false,
  "shows": Array [
    Object {
      "id": 1,
      "name": "Breaking Bad",
    },
    Object {
      "id": 2,
      "name": "Friends",
    },
  ],
}
`;

exports[`useShows composable with snapshots fetches shows successfully: loading state 1`] = `
Object {
  "error": null,
  "loading": true,
  "shows": Array [],
}
`;

exports[`useShows composable with snapshots handles API errors correctly 1`] = `
Object {
  "error": "API Error Message",
  "loading": false,
  "shows": Array [],
}
`;

exports[`useShows composable with snapshots handles unknown errors gracefully 1`] = `
Object {
  "error": "Network issue",
}
`;

exports[`useShows composable with snapshots initializes with default values 1`] = `
Object {
  "error": null,
  "loading": false,
  "shows": Array [],
}
`;
